colexec: request file descriptors up front in external sorts/joins

asubiotto opened the pull request:
Release justification: fix for high-severity bug in existing functionality. The
current methodology of requesting file descriptors one at a time could lead to
a potential deadlock when spilling many queries to disk at the same time.

These operators would previously call semaphore.Acquire as needed. This could
lead to a situation where concurrent sorts/joins could block since they could
potentially deadlock Acquiring a file descriptor and never release ones
already held.

Now, sorts and joins that spill to disk will acquire all necessary resources
before spilling to disk.

Release note: None (no release with this potential issue)

Fixes #45602

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45892)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2-rCLFCVr6H9lE4Jkw:-M2-rCLFCVr6H9lE4Jkx:bkxexv) ([raw file](https://github.com/cockroachdb/cockroach/blob/49eb612ac320c9c57baae08f0e9708fcc28124be/pkg/sql/colexec/external_hash_joiner.go#L266)):*
> ```Go
> 		}
> 	}
> 	if maxNumberActivePartitions < 4 {
> ```

We should probably now have a hard limit on `maxNumberActivePartitions` dependent on `VecMaxOpenFDsLimit` (maybe 1/8 of it?).
___
*[pkg/sql/colexec/external_sort.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2-quHn3xktSmNqwjM-:-M2-quHn3xktSmNqwjM0:b-mpm956) ([raw file](https://github.com/cockroachdb/cockroach/blob/49eb612ac320c9c57baae08f0e9708fcc28124be/pkg/sql/colexec/external_sort.go#L196)):*
> ```Go
> 	// In order to make progress when merging we have to merge at least two
> 	// partitions.
> 	if maxNumberPartitions < 2 {
> ```

ditto for `maxNumberPartitions`.


<!-- Sent from Reviewable.io -->

asubiotto commented:
I'll wait for #45825 to go in before doing some more work on this one.

bobvawter commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=45892) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=45892) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=45892) it.</sub>

asubiotto commented:
Rebased and updated to limit the number of file descriptors a sorter requests. I'm slightly confused with our current external hash join partition limits and minimums, which I left as a TODO to discuss.

yuzefovich reviewed (commented):
> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2-rCLFCVr6H9lE4Jkw:-M2Et-Qc9HZQkpBV5MN8:bvagxa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/49eb612ac320c9c57baae08f0e9708fcc28124be/pkg/sql/colexec/external_hash_joiner.go#L266)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

We should probably now have a hard limit on `maxNumberActivePartitions` dependent on `VecMaxOpenFDsLimit` (maybe 1/8 of it?).
</blockquote></details>

Do you want to address this suggestion in this PR or in the follow-up?
___
*[pkg/sql/colexec/external_hash_joiner.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2ErY342lZ5fBHsynKD:-M2ErY342lZ5fBHsynKE:bl81fyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L104)):*
> ```Go
> 	// fallback to sort and merge join. We'll be using the minimum necessary per
> 	// input + 4 (2 for each spilling queue that the merge joiner uses).
> 	// TODO(asubiotto): Is this right?
> ```

Yes, I believe the number is correct. I'd probably add a more verbose description of this computation, for example:
```
During the fallback, at minimum we can have the following:
- 2 FDs will be used to read from the partition of the external hash joiner that could not have been repartitioned, these are inputs to the externalSorters
- 2 * externalSorterMinPartitions FDs will be used by two externalSorters that are inputs into the merge joiner
- 2 FDs will be used by the merge joiner
```
___
*[pkg/sql/colexec/external_hash_joiner.go, line 217 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EsXaH5J9UKQCbCqZC:-M2EsXaIA5eJyiejQ1Zw:ba3vw4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L217)):*
> ```Go
> 		// needed) before it proceeds to actual join partitions.
> 		numForcedRepartitions int
> 		// delegateFDAcquisitions if true, means that a test wants to force the
> ```

nit: `s/ if true,/, if true,/g`.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EshLQDdVOugBaqm2I:-M2EshLQDdVOugBaqm2J:b-c6b4ui) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L249)):*
> ```Go
> // - delegateFDAcquisitions specifies whether the external hash joiner should
> // let the partitioned disk queues acquire file descriptors instead of acquiring
> // them up front in Next.
> ```

nit: I'd add that this should be `true` only in tests.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtLJWCXg2qAH0Bqrn:-M2EtLJWCXg2qAH0Bqro:b7o5ffn) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L311)):*
> ```Go
> 	externalSorterMaxNumberPartitions := (maxNumberActivePartitions - 4) / 2
> 	if externalSorterMaxNumberPartitions < externalSorterMinPartitions {
> 		externalSorterMaxNumberPartitions = externalSorterMinPartitions
> ```

I feel like we should panic in this case because we're upping `maxNumberActivePartitions` so that there is a least minimum number of FDs needed for each of the external sorters.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtZdhCcylijzEKPwk:-M2EtZdhCcylijzEKPwl:brq5gzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L335)):*
> ```Go
> 		execerror.VectorizedInternalPanic(err)
> 	}
> 	fmt.Println("max number of active partitions is", maxNumberActivePartitions)
> ```

Leftover logging.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 702 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EthN56oNxO64shCXf:-M2EthN56oNxO64shCXg:bnif7rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L702)):*
> ```Go
> 	}
> 	if !hj.testingKnobs.delegateFDAcquisitions && hj.fdState.acquiredFDs > 0 {
> 		hj.fdState.fdSemaphore.Release(hj.fdState.acquiredFDs)
> ```

nit: for completeness, we probably should do `hj.fdState.acquiredFDs = 0`.
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2Ev0Ky6SvmYPil3gXy:-M2Ev0Ky6SvmYPil3gXz:brqm0d9) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner_test.go#L43)):*
> ```Go
> // the fallback to sort + merge join we will be using 6 (external sorter on
> // each side will use 2 to read its own partitions and 1 more to write out).
> // TODO(asubiotto): It would probably be good to discuss this again.
> ```

I think this comment is incorrect. I now believe that we need, in theory, the minimum should be 10. I think we should remove this constant and use `externalHJMinPartitions` in the tests.
___
*[pkg/sql/colexec/external_sort.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvLjR8Smy6MPOIBVf:-M2EvLjR8Smy6MPOIBVg:b-40ic7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L137)):*
> ```Go
> 
> 	testingKnobs struct {
> 		// delegateFDAcquisitions if true, means that a test wants to force the
> ```

nit: `s/ if/, if/g`.
___
*[pkg/sql/colexec/external_sort.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvS917ywU4CImeYQR:-M2EvS917ywU4CImeYQS:bu6iph4) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L163)):*
> ```Go
> // - delegateFDAcquisitions specifies whether the external sorter should let
> // the partitioned disk queue acquire file descriptors instead of acquiring
> // them up front in Next.
> ```

ditto for the "tests only".
___
*[pkg/sql/colexec/external_sort.go, line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EveXGDl2CRQ1jEMtd:-M2EveXGDl2CRQ1jEMte:brq5gzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L198)):*
> ```Go
> 		// external sorter
> 		// TODO(asubiotto): this number should be tuned.
> 		fmt.Println("maxNumberPartitions", maxNumberPartitions)
> ```

Leftover logging.
___
*[pkg/sql/colexec/external_sort.go, line 199 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvftRDbmVrUdx9jcl:-M2EvftRDbmVrUdx9jcm:bxklxq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L199)):*
> ```Go
> 		// TODO(asubiotto): this number should be tuned.
> 		fmt.Println("maxNumberPartitions", maxNumberPartitions)
> 		maxPartitionsAllowed := fdSemaphore.GetLimit() / 16
> ```

nit: we probably should now extract this computation into a function used by both sorter and joiner.
___
*[pkg/sql/colexec/external_sort.go, line 365 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2Evr-i6IeZ0FGHg4ky:-M2Evr-i6IeZ0FGHg4kz:b-1bqqs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L365)):*
> ```Go
> 	}
> 	if !s.testingKnobs.delegateFDAcquisitions && s.fdState.fdSemaphore != nil && s.fdState.acquiredFDs > 0 {
> 		s.fdState.fdSemaphore.Release(s.fdState.acquiredFDs)
> ```

ditto to zero out `.acquiredFDs`.
___
*[pkg/sql/colexec/external_sort_test.go, line 152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvwjT7hQE2z43COUS:-M2EvwjT7hQE2z43COUT:b-g2mlrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort_test.go#L152)):*
> ```Go
> 		memMonitors []*mon.BytesMonitor
> 	)
> 	const maxNumberPartitions = 3
> ```

nit: these constants should probably be removed in favor of `externalSorterMinPartitions`.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2-rCLFCVr6H9lE4Jkw:-M2IjvcA1fCuuGt8OMbC:b-vhn1kh) ([raw file](https://github.com/cockroachdb/cockroach/blob/49eb612ac320c9c57baae08f0e9708fcc28124be/pkg/sql/colexec/external_hash_joiner.go#L266)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do you want to address this suggestion in this PR or in the follow-up?
</blockquote></details>

`fdSemaphore.GetLimit()/16` above is what sets the maximum. I think this is fine, no?
___
*[pkg/sql/colexec/external_hash_joiner.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2ErY342lZ5fBHsynKD:-M2Ihhug0GlItbQjtipd:byjszq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L104)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yes, I believe the number is correct. I'd probably add a more verbose description of this computation, for example:
```
During the fallback, at minimum we can have the following:
- 2 FDs will be used to read from the partition of the external hash joiner that could not have been repartitioned, these are inputs to the externalSorters
- 2 * externalSorterMinPartitions FDs will be used by two externalSorters that are inputs into the merge joiner
- 2 FDs will be used by the merge joiner
```
</blockquote></details>

Done. It seems like the correct number, but I think the merge joiner will use 4 FDs since it uses 2 spilling queues, right?
___
*[pkg/sql/colexec/external_hash_joiner.go, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtLJWCXg2qAH0Bqrn:-M2IkFxf1s-1Vt41fFVa:bay4ec4) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L311)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I feel like we should panic in this case because we're upping `maxNumberActivePartitions` so that there is a least minimum number of FDs needed for each of the external sorters.
</blockquote></details>

Hmm, this doesn't go hand in hand with using the semaphore limit to set a maximum on the number of active partitions, since tests will set this limit low to ensure the hash joiner doesn't exceed the limit. This results in passing in a limit of 0 to the sorter with a nil `fdSemaphore`. I don't do a nil check there because we should never pass in a limit of 0 to a sorter if we account for its maximum number of partitions. In practice, this if statement only happens in tests, so I'm fine leaving as is if you are.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtZdhCcylijzEKPwk:-M2IvCym3nQ0ogINlyAR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L335)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Leftover logging.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2Ev0Ky6SvmYPil3gXy:-M2IxKB6A3jBZvhNsUOk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner_test.go#L43)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this comment is incorrect. I now believe that we need, in theory, the minimum should be 10. I think we should remove this constant and use `externalHJMinPartitions` in the tests.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/external_sort.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2-quHn3xktSmNqwjM-:-M2Iyx0l5DzIZr__sFvL:b4cdl12) ([raw file](https://github.com/cockroachdb/cockroach/blob/49eb612ac320c9c57baae08f0e9708fcc28124be/pkg/sql/colexec/external_sort.go#L196)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for `maxNumberPartitions`.
</blockquote></details>

Replied there.
___
*[pkg/sql/colexec/external_sort.go, line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EveXGDl2CRQ1jEMtd:-M2IxXyp-jkuaxXSImG7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L198)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Leftover logging.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/external_sort.go, line 199 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvftRDbmVrUdx9jcl:-M2IxRiC7I-1mMwJ4-yy:b-6fhsnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L199)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: we probably should now extract this computation into a function used by both sorter and joiner.
</blockquote></details>

I don't think it's worth it. If we tune this, the numbers will probably be different so we'll have to change this anyway.
___
*[pkg/sql/colexec/external_sort.go, line 365 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2Evr-i6IeZ0FGHg4ky:-M2IyycMB7-XRe0-MJ8I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L365)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto to zero out `.acquiredFDs`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2-rCLFCVr6H9lE4Jkw:-M2JIes46Rsx78f8QD2-:b-b61iji) ([raw file](https://github.com/cockroachdb/cockroach/blob/49eb612ac320c9c57baae08f0e9708fcc28124be/pkg/sql/colexec/external_hash_joiner.go#L266)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`fdSemaphore.GetLimit()/16` above is what sets the maximum. I think this is fine, no?
</blockquote></details>

Oh yeah, I missed that.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2ErY342lZ5fBHsynKD:-M2JB9omEW_HAOT9_t9T:b-u0yae9) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L104)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done. It seems like the correct number, but I think the merge joiner will use 4 FDs since it uses 2 spilling queues, right?
</blockquote></details>

Merge joiner follows the pattern of "write all data, then read all data", so each queue should be using 1 FD at any point. (Probably merge joiner should use "close on new partition" strategy.) However, as I said in the comment above, we should not be forgetting about the FD to read from the partition that we are joining (but it is outside of merge joiner's knowledge that we use a FD there).
___
*[pkg/sql/colexec/external_hash_joiner.go, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtLJWCXg2qAH0Bqrn:-M2JJ0id5yGBXokJZjzB:b41cwke) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L311)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Hmm, this doesn't go hand in hand with using the semaphore limit to set a maximum on the number of active partitions, since tests will set this limit low to ensure the hash joiner doesn't exceed the limit. This results in passing in a limit of 0 to the sorter with a nil `fdSemaphore`. I don't do a nil check there because we should never pass in a limit of 0 to a sorter if we account for its maximum number of partitions. In practice, this if statement only happens in tests, so I'm fine leaving as is if you are.
</blockquote></details>

SGTM. I'd add a comment here then that this `if` is in-place to handle artificially low FDs limits set by tests.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtZdhCcylijzEKPwk:-M2JJCq_2g2Q2p7-D5p9:b-bsn059) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L335)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I still see it :)
___
*[pkg/sql/colexec/external_hash_joiner.go, line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JDgUWChpI_lspy_Ly:-M2JDgUWChpI_lspy_Lz:b-sinbr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c571cb17e6d3f45dcacd04d024e77947539b8d1/pkg/sql/colexec/external_hash_joiner.go#L109)):*
> ```Go
> 	//   sort and partition this input. At this stage 2 + 2 *
> 	//   externalMinPartitions FDs are used.
> 	// - Once the inputs (the hash joiner partitions) are finished, both FDs will
> ```

I think it is possible that the hash joiner partitions' FDs are not released when the merge joiner kicks in. Consider the following scenario (looking only at one side):
- merge joiner requests the first batch from the external sorter (when it gets the batch, it will be written into the spilling queue) + 1 FD
- external sorter requests batches from the hash joiner partition and does its job. At some point, however, the external sorter needed to do recursive merging, and at that point it was using all of its available FDs. + `externalSorterMinPartitions` FDs
- the hash joiner partition is being read + 1 FD.

However, while typing out this comment, I realized that the merge joiner will start writing into its spilling queue only when it receives the first batch from the external sorter, and at that point, the hash joiner partition has been fully consumed, and the corresponding FD should have been released. I.e. we actually never use more than `1 + externalSorterMinPartitions` number of FDs per side, so `externalHJMinPartitions` should be `2 + 2 * externalSorterMinPartitions`.
___
*[pkg/sql/colexec/external_sort.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvS917ywU4CImeYQR:-M2JKfQT520SkbdmqnXF:b-82oqa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L163)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for the "tests only".
</blockquote></details>

Ping.
___
*[pkg/sql/colexec/external_sort.go, line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EveXGDl2CRQ1jEMtd:-M2JKdVDCWuGDNJCWJOJ:b-wg7n1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L198)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still see it.
___
*[pkg/sql/colexec/external_sort.go, line 199 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvftRDbmVrUdx9jcl:-M2JKbWf0PQZAJdffbNj:b2nhll9) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L199)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think it's worth it. If we tune this, the numbers will probably be different so we'll have to change this anyway.
</blockquote></details>

Ok.
___
*[pkg/sql/colexec/external_sort.go, line 216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JAp1-2NjwLnGJXnxS:-M2JAp1-2NjwLnGJXnxT:b-gc35d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c571cb17e6d3f45dcacd04d024e77947539b8d1/pkg/sql/colexec/external_sort.go#L216)):*
> ```Go
> 		OneInputNode:       NewOneInputNode(inMemSorter),
> 		unlimitedAllocator: unlimitedAllocator,
> 		memoryLimit:        memoryLimit,
> ```

nit: external sorter doesn't need to store `memoryLimit`.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2ErY342lZ5fBHsynKD:-M2JVojK3BPZ-VAXSpyO:bdfqym8) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L104)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Merge joiner follows the pattern of "write all data, then read all data", so each queue should be using 1 FD at any point. (Probably merge joiner should use "close on new partition" strategy.) However, as I said in the comment above, we should not be forgetting about the FD to read from the partition that we are joining (but it is outside of merge joiner's knowledge that we use a FD there).
</blockquote></details>

Just checking, is this comment void now after the `However, while typing out this comment`? I responded re the `1FD` above, it uses only 1, but actually acquires `2` right now (since there is no enforcing of the fact that all writes happen before all reads in the spilling queue, we should probably fix that)
___
*[pkg/sql/colexec/external_hash_joiner.go, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtLJWCXg2qAH0Bqrn:-M2JXD6iC_qIJgyi2A-_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L311)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

SGTM. I'd add a comment here then that this `if` is in-place to handle artificially low FDs limits set by tests.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EtZdhCcylijzEKPwk:-M2JYzos4Ynmc_Cw2mHq:bidxqgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L335)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I still see it :)
</blockquote></details>

Oops, thanks for catching
___
*[pkg/sql/colexec/external_hash_joiner.go, line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JDgUWChpI_lspy_Ly:-M2JV6Ho96PZxnaoBJkM:b-96xujc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c571cb17e6d3f45dcacd04d024e77947539b8d1/pkg/sql/colexec/external_hash_joiner.go#L109)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think it is possible that the hash joiner partitions' FDs are not released when the merge joiner kicks in. Consider the following scenario (looking only at one side):
- merge joiner requests the first batch from the external sorter (when it gets the batch, it will be written into the spilling queue) + 1 FD
- external sorter requests batches from the hash joiner partition and does its job. At some point, however, the external sorter needed to do recursive merging, and at that point it was using all of its available FDs. + `externalSorterMinPartitions` FDs
- the hash joiner partition is being read + 1 FD.

However, while typing out this comment, I realized that the merge joiner will start writing into its spilling queue only when it receives the first batch from the external sorter, and at that point, the hash joiner partition has been fully consumed, and the corresponding FD should have been released. I.e. we actually never use more than `1 + externalSorterMinPartitions` number of FDs per side, so `externalHJMinPartitions` should be `2 + 2 * externalSorterMinPartitions`.
</blockquote></details>

Right regarding that the FD is released. But note that each spilling queue, even though it only opens 1 FD, `Acquire`s 2 of them (which is what needs to be reflected here), which is why it's `4 + 2 * externalSorterMinPartitions`.
___
*[pkg/sql/colexec/external_sort.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EvS917ywU4CImeYQR:-M2JZ7YY7841wGU7cQg_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L163)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Ping.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/external_sort.go, line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2EveXGDl2CRQ1jEMtd:-M2JZGuOCHIVEVVs6kak:bg8t3q0) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_sort.go#L198)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Still see it.
</blockquote></details>

I don't think it's there anymore.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2ErY342lZ5fBHsynKD:-M2Ja3VHDErzesTzFNys:b-8775au) ([raw file](https://github.com/cockroachdb/cockroach/blob/30958a14e2668f35084469df6b363760c8ffcc09/pkg/sql/colexec/external_hash_joiner.go#L104)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Just checking, is this comment void now after the `However, while typing out this comment`? I responded re the `1FD` above, it uses only 1, but actually acquires `2` right now (since there is no enforcing of the fact that all writes happen before all reads in the spilling queue, we should probably fix that)
</blockquote></details>

Yeah, resolved.
___
*[pkg/sql/colexec/external_hash_joiner.go, line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JDgUWChpI_lspy_Ly:-M2J_TGc3shd7CxOslrE:b-g3yyrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c571cb17e6d3f45dcacd04d024e77947539b8d1/pkg/sql/colexec/external_hash_joiner.go#L109)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Right regarding that the FD is released. But note that each spilling queue, even though it only opens 1 FD, `Acquire`s 2 of them (which is what needs to be reflected here), which is why it's `4 + 2 * externalSorterMinPartitions`.
</blockquote></details>

Oh, I see, I forgot about that. I guess we could introduce non-default mode for spilling queue so that it would acquire only 1 FD. I'll leave it up to you to decide whether it is important to add a TODO for this or we can just ignore this.
___
*[pkg/sql/colexec/external_sort.go, line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JaPmo5B5QA2OCjlaj:-M2JaPmo5B5QA2OCjlak:b-xcbjtg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d8ca3acc9f8c972362a81ed4f0741978b65dcd/pkg/sql/colexec/external_sort.go#L198)):*
> ```Go
> 	}
> 	// Each disk queue will use up to BufferSizeBytes of RAM, so we reduce the
> 	// memoryLimit of the partitions to sort in memory by those cache sizes.will give
> ```

nit: "sizes.will"
___
*[pkg/sql/colexec/external_sort.go, line 203 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JanQS8BPmH0n6zyX9:-M2JanQS8BPmH0n6zyXA:bsv72xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d8ca3acc9f8c972362a81ed4f0741978b65dcd/pkg/sql/colexec/external_sort.go#L203)):*
> ```Go
> 	batchMemSize := estimateBatchSizeBytes(inputTypes, coldata.BatchSize())
> 	// Reserve a certain amount of memory for the partition caches.
> 	memoryLimit -= int64((maxNumberPartitions * diskQueueCfg.BufferSizeBytes) + batchMemSize)
> ```

At this point this update no longer matters. We need to move the creation of `inputPartitioningOperator` after this update.
___
*[pkg/sql/colexec/external_sort.go, line 205 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JatBZ9q4_mt1l_Ant:-M2JatBZ9q4_mt1l_Anu:b9f458n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d8ca3acc9f8c972362a81ed4f0741978b65dcd/pkg/sql/colexec/external_sort.go#L205)):*
> ```Go
> 	memoryLimit -= int64((maxNumberPartitions * diskQueueCfg.BufferSizeBytes) + batchMemSize)
> 	if memoryLimit < 0 {
> 		memoryLimit = 0
> ```

I think that we need to have memory limit of at least 1 byte (otherwise `inputPartitioningOperator` will only return zero-length batches, always).


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JDgUWChpI_lspy_Ly:-M2Jjm7e5TR0T_nEm8SU:bjuc9tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c571cb17e6d3f45dcacd04d024e77947539b8d1/pkg/sql/colexec/external_hash_joiner.go#L109)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Oh, I see, I forgot about that. I guess we could introduce non-default mode for spilling queue so that it would acquire only 1 FD. I'll leave it up to you to decide whether it is important to add a TODO for this or we can just ignore this.
</blockquote></details>

It was actually very easy to add (just check for whether the queue is rewindable or not) so added it.
___
*[pkg/sql/colexec/external_sort.go, line 203 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JanQS8BPmH0n6zyX9:-M2Jqnxw7GRLIKPHDN8c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d8ca3acc9f8c972362a81ed4f0741978b65dcd/pkg/sql/colexec/external_sort.go#L203)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

At this point this update no longer matters. We need to move the creation of `inputPartitioningOperator` after this update.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/external_sort.go, line 205 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2JatBZ9q4_mt1l_Ant:-M2Jr6AeCga-4iPs_TTJ:b-cqt3vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d8ca3acc9f8c972362a81ed4f0741978b65dcd/pkg/sql/colexec/external_sort.go#L205)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think that we need to have memory limit of at least 1 byte (otherwise `inputPartitioningOperator` will only return zero-length batches, always).
</blockquote></details>

Done. This makes us hit what #46052 fixes a lot more easily.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/spilling_queue.go, line 210 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2K1rDU-aWsvzXdbsV9:-M2K1rDV7Zo0AE9ZFwaF:b-4oepwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9384eb48d33b9fc3ffbf34475a55bb75e8ddbaa2/pkg/sql/colexec/spilling_queue.go#L210)):*
> ```Go
> 
> func (q *spillingQueue) numFDsOpenAtAnyGivenTime() int {
> 	if q.rewindable {
> ```

Merge joiner uses rewindable queue only on the right side, yet it follows the pattern "write all, then read all" on both sides. We probably need a separate knob for this pattern after all. (Or, alternatively, we could be instantiating a rewindable queue in the merge joiner on both sides, but using `rewind()` method only on the right.)


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/spilling_queue.go, line 210 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2K1rDU-aWsvzXdbsV9:-M2KDMHbCizAxLG-T26E:bei84io) ([raw file](https://github.com/cockroachdb/cockroach/blob/9384eb48d33b9fc3ffbf34475a55bb75e8ddbaa2/pkg/sql/colexec/spilling_queue.go#L210)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Merge joiner uses rewindable queue only on the right side, yet it follows the pattern "write all, then read all" on both sides. We probably need a separate knob for this pattern after all. (Or, alternatively, we could be instantiating a rewindable queue in the merge joiner on both sides, but using `rewind()` method only on the right.)
</blockquote></details>

The check is easy enough. If the underlying disk queue is not using a default cache mode then this pattern is enforced, so we can use that.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-:-M2KEMtt-SCYKNISX21W:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2KEpp40OocMNR-Ikwf:-M2KEpp40OocMNR-Ikwg:b9d4eh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c3e54ae2759d55b9fb7c6e6850266c2c8cc60d8/pkg/sql/colexec/external_hash_joiner.go#L320)):*
> ```Go
> 	// the partitions that we need to join using sort + merge join strategy, and
> 	// all others are divided between the two inputs.
> 	externalSorterMaxNumberPartitions := (maxNumberActivePartitions - 4) / 2
> ```

`s/4/2/g`.

Maybe we should define a separate constant for this `2` so that we don't forget to update one of the places where this constant is used if the calculation changes for some reason?
___
*[pkg/sql/colexec/external_sort_test.go, line 89 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2KFnNu2FQRyKDwn5MY:-M2KFnNu2FQRyKDwn5MZ:b-w1mgv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c3e54ae2759d55b9fb7c6e6850266c2c8cc60d8/pkg/sql/colexec/external_sort_test.go#L89)):*
> ```Go
> 						orderedVerifier,
> 						func(input []Operator) (Operator, error) {
> 							// A sorter should never exceed maxNumberPartitions+1, even during
> ```

nit: the comment needs updating.
___
*[pkg/sql/colexec/spilling_queue.go, line 210 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2K1rDU-aWsvzXdbsV9:-M2KECFX266S1PrEcN17:b-81nlyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9384eb48d33b9fc3ffbf34475a55bb75e8ddbaa2/pkg/sql/colexec/spilling_queue.go#L210)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The check is easy enough. If the underlying disk queue is not using a default cache mode then this pattern is enforced, so we can use that.
</blockquote></details>

SGTM.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45892)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/external_hash_joiner.go, line 320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45892#-M2KEpp40OocMNR-Ikwf:-M2Kd6049vLtthsyemhx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c3e54ae2759d55b9fb7c6e6850266c2c8cc60d8/pkg/sql/colexec/external_hash_joiner.go#L320)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

`s/4/2/g`.

Maybe we should define a separate constant for this `2` so that we don't forget to update one of the places where this constant is used if the calculation changes for some reason?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto commented:
bors r=yuzefovich

Commit:
colexec: request file descriptors up front in external sorts/joins

Release justification: fix for high-severity bug in existing functionality. The
current methodology of requesting file descriptors one at a time could lead to
a potential deadlock when spilling many queries to disk at the same time.

These operators would previously call semaphore.Acquire as needed. This could
lead to a situation where concurrent sorts/joins could block since they could
potentially deadlock Acquiring a file descriptor and never release ones
already held.

Now, sorts and joins that spill to disk will acquire all necessary resources
before spilling to disk.

Release note: None (no release with this potential issue)

asubiotto commented:
A release justification was needed in the PR description as well. Updated.

bors r=yuzefovich
